File: android/core/src/main/java/io/github/bjthomas07/pushdispatch/core/NotificationModels.kt
package io.github.bjthomas07.pushdispatch.core

enum class NotificationProvider(val wireValue: String) {
    FCM("fcm"),
}

enum class NotificationTargetType(val wireValue: String) {
    FIREBASE_INSTALLATION_ID("fid"),
    REGISTRATION_TOKEN("token"),
}

data class NotificationTarget(
    val provider: NotificationProvider,
    val type: NotificationTargetType,
    val value: String,
) {
    init {
        require(value.isNotBlank()) { "Notification target must not be blank" }
    }
}

enum class NotificationPermission(val wireValue: String) {
    GRANTED("granted"),
    DENIED("denied"),
    UNKNOWN("unknown"),
}

data class NotificationInstallationMetadata(
    val platform: String,
    val permission: NotificationPermission,
    val timezone: String,
    val enabled: Boolean,
) {
    init {
        require(platform.isNotBlank()) { "Notification platform must not be blank" }
        require(timezone.isNotBlank()) { "Notification timezone must not be blank" }
    }
}

data class NotificationInstallationState(
    /** Local-only account key. Authentication on the HTTP request performs the actual binding. */
    val accountKey: String,
    val installationId: String,
    val target: NotificationTarget,
    val metadata: NotificationInstallationMetadata,
    val lastSeenAtEpochMillis: Long,
    /** Installation IDs that were replaced but whose backend cleanup still needs a retry. */
    val pendingCleanupInstallationIds: List<String> = emptyList(),
)

data class NotificationMessage(
    val messageId: String?,
    val title: String?,
    val body: String?,
    val deepLink: String?,
    val data: Map<String, String>,
    val collapseKey: String? = null,
)

fun interface NotificationTargetProvider {
    suspend fun register(): NotificationTarget
}

class NotificationAccountTransitionException : IllegalStateException(
    "Notification installation is still bound to another account; unbind must succeed first",
)

/** Backend adapters throw this when heartbeat finds no stored installation. */
class NotificationInstallationNotFoundException(
    cause: Throwable? = null,
) : Exception("Notification installation no longer exists", cause)

interface NotificationInstallationBackend {
    /** Creates or re-binds the target to the account represented by the current auth context. */
    suspend fun bind(
        target: NotificationTarget,
        metadata: NotificationInstallationMetadata,
    ): String

    suspend fun heartbeat(
        installationId: String,
        metadata: NotificationInstallationMetadata,
    )

    suspend fun unbind(installationId: String)
}

interface NotificationInstallationStateStore {
    fun read(): NotificationInstallationState?
    fun write(state: NotificationInstallationState)
    fun clear()
}

fun interface NotificationClock {
    fun nowEpochMillis(): Long
}

File: ios/Sources/PushDispatchCore/NotificationModels.swift
import Foundation

public enum NotificationProvider: String, Codable, Sendable {
    case fcm
}

public enum NotificationTargetType: String, Codable, Sendable {
    case firebaseInstallationID = "fid"
    case registrationToken = "token"
}

public enum NotificationPlatform: String, Codable, Sendable {
    case iOS = "ios"
    case android
}

/// Backend-facing permission state. Provisional and ephemeral iOS permission
/// both map to `granted`; unsupported/future states map to `unknown`.
public enum NotificationPermission: String, Codable, Sendable {
    case granted
    case denied
    case unknown
}

public struct NotificationTarget: Codable, Equatable, Sendable {
    public let provider: NotificationProvider
    public let type: NotificationTargetType
    public let value: String

    public init(
        provider: NotificationProvider,
        type: NotificationTargetType,
        value: String
    ) {
        self.provider = provider
        self.type = type
        self.value = value
    }
}

public struct NotificationInstallationMetadata: Codable, Equatable, Sendable {
    public let platform: NotificationPlatform
    public let permission: NotificationPermission
    public let timezone: String
    public let enabled: Bool

    public init(
        platform: NotificationPlatform,
        permission: NotificationPermission,
        timezone: String,
        enabled: Bool = true
    ) {
        self.platform = platform
        self.permission = permission
        self.timezone = timezone
        self.enabled = enabled
    }
}

/// Authentication identifies the user; callers cannot supply a UID in the body.
public struct NotificationRegistrationRequest: Codable, Equatable, Sendable {
    public let provider: NotificationProvider
    public let targetType: NotificationTargetType
    public let target: String
    public let platform: NotificationPlatform
    public let permission: NotificationPermission
    public let timezone: String
    public let enabled: Bool

    public init(
        target: NotificationTarget,
        platform: NotificationPlatform,
        permission: NotificationPermission,
        timezone: String,
        enabled: Bool = true
    ) {
        self.provider = target.provider
        self.targetType = target.type
        self.target = target.value
        self.platform = platform
        self.permission = permission
        self.timezone = timezone
        self.enabled = enabled
    }

    public var metadata: NotificationInstallationMetadata {
        NotificationInstallationMetadata(
            platform: platform,
            permission: permission,
            timezone: timezone,
            enabled: enabled
        )
    }
}

public struct NotificationHeartbeatRequest: Codable, Equatable, Sendable {
    public let permission: NotificationPermission?
    public let timezone: String?
    public let enabled: Bool?

    public init(
        permission: NotificationPermission? = nil,
        timezone: String? = nil,
        enabled: Bool? = nil
    ) {
        self.permission = permission
        self.timezone = timezone
        self.enabled = enabled
    }
}

public struct NotificationInstallationReceipt: Codable, Equatable, Sendable {
    public let installationID: String

    public init(installationID: String) {
        self.installationID = installationID
    }

    private enum CodingKeys: String, CodingKey {
        case installationID = "installationId"
    }
}

public enum NotificationInstallationTransportError: Error, Equatable, Sendable {
    case installationNotFound
}

/// App-owned authenticated transport for installation lifecycle requests.
public protocol NotificationInstallationTransport: Sendable {
    func register(
        _ request: NotificationRegistrationRequest
    ) async throws -> NotificationInstallationReceipt

    func heartbeat(
        installationID: String,
        request: NotificationHeartbeatRequest
    ) async throws -> NotificationInstallationReceipt

    func unregister(installationID: String) async throws
}
